packages/safe-ds-vscode/src/extension/mainClient.ts

```diff
@@ -137,12 +141,177 @@ const registerMessageLoggingCallbacks = function () {
+                            });
+                        },
+                    );
+                    const cleanupLoadingIndication = () => {
+                        loadingInProgress = false;
+                    };
+
+                    const placeHolderTypeCallback = function (message: PlaceholderTypeMessage) {
+                        printOutputMessage(
+                            `Placeholder was calculated (${message.id}): ${message.data.name} of type ${message.data.type}`,
+                        );
+                        if (
+                            message.id === pipelineId &&
+                            message.data.type === 'Table' &&
+                            message.data.name === requestedPlaceholderName
+                        ) {
+                            lastFinishedPipelineId = pipelineId;
+                            lastSuccessfulPlaceholderName = requestedPlaceholderName;
+                            EDAPanel.createOrShow(
+                                context.extensionUri,
+                                context,
+                                pipelineId,
+                                message.data.name,
+                                pythonServerPort,
+                            );
+                            removeMessageCallback(placeHolderTypeCallback, 'placeholder_type');
+                            cleanupLoadingIndication();
+                        } else if (message.id === pipelineId && message.data.name !== requestedPlaceholderName) {
+                            return;
+                        } else if (message.id === pipelineId) {
+                            lastFinishedPipelineId = pipelineId;
+                            vscode.window.showErrorMessage(`Selected placeholder is not of type 'Table'.`);
+                            removeMessageCallback(placeHolderTypeCallback, 'placeholder_type');
+                            cleanupLoadingIndication();
+                        }
+                    };
+                    addMessageCallback(placeHolderTypeCallback, 'placeholder_type');
+
+                    const runtimeProgressCallback = function (message: RuntimeProgressMessage) {
+                        printOutputMessage(`Runner-Progress (${message.id}): ${message.data}`);
+                        if (
+                            message.id === pipelineId &&
+                            message.data === 'done' &&
+                            lastFinishedPipelineId !== pipelineId
+                        ) {
+                            lastFinishedPipelineId = pipelineId;
+                            vscode.window.showErrorMessage(`Selected text is not a placeholder!`);
+                            removeMessageCallback(runtimeProgressCallback, 'runtime_progress');
+                            cleanupLoadingIndication();
+                        }
+                    };
+                    addMessageCallback(runtimeProgressCallback, 'runtime_progress');
```

Comment:
Just an aside: If the parameters of `addMessageCallback` were switched, it be quite readable to inline the callback:

```
addMessageCallback('runtime_progress', (message: RuntimeProgressMessage) => {
    // ...
});